lib/client/debug/debug.go

```diff
@@ -0,0 +1,152 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package debug
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/trace"
+
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+)
+
+// Client specifies the debug service client.
+type Client interface {
+	// SetLogLevel changes the application's log level and a change status message.
+	SetLogLevel(context.Context, string) (string, error)
+	// GetLogLevel fetches the current log level.
+	GetLogLevel(context.Context) (string, error)
+	// CollectProfile collects a pprof profile.
+	CollectProfile(context.Context, string, int) ([]byte, error)
+}
+
+type clt struct {
+	clt *http.Client
+}
+
+// NewClient generates a new debug service client.
+func NewClient(socketPath string) Client {
+	return &clt{
+		clt: &http.Client{
+			Timeout: apidefaults.DefaultIOTimeout,
+			Transport: &http.Transport{
+				DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+					return net.Dial("unix", socketPath)
+				},
+			},
+		},
+	}
+}
+
+// SetLogLevel changes the application's log level and a change status message.
+func (c *clt) SetLogLevel(ctx context.Context, level string) (string, error) {
+	resp, err := c.do(ctx, http.MethodPut, url.URL{Path: teleport.DebugLogLevelEndpoint}, []byte(level))
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	respBody, err := io.ReadAll(io.LimitReader(resp.Body, 1024))
+	defer resp.Body.Close()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return "", trace.BadParameter("Unable to change log level: %s", respBody)
+	}
+
+	return string(respBody), nil
+}
+
+// GetLogLevel fetches the current log level.
+func (c *clt) GetLogLevel(ctx context.Context) (string, error) {
+	resp, err := c.do(ctx, http.MethodGet, url.URL{Path: teleport.DebugLogLevelEndpoint}, nil)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	respBody, err := io.ReadAll(io.LimitReader(resp.Body, 1024))
+	defer resp.Body.Close()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return "", trace.BadParameter("Unable to fetch log level: %s", respBody)
+	}
+
+	return string(respBody), nil
+}
+
+// CollectProfile collects a pprof profile.
+func (c *clt) CollectProfile(ctx context.Context, profileName string, seconds int) ([]byte, error) {
+	u := url.URL{
+		Path: teleport.DebugPProfEndpointsPrefix + profileName,
+	}
+
+	if seconds > 0 {
+		qs := url.Values{}
+		qs.Add("seconds", strconv.Itoa(seconds))
+		u.RawQuery = qs.Encode()
+	}
+
+	resp, err := c.do(ctx, http.MethodGet, u, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	result, err := io.ReadAll(resp.Body)
```

Comment:
We can add the limit, but not returning an error when the limit is reached might not be the best approach (because it will cause them to be broken). Given that this is related to how much of the application is being used/its size, it won't be easy to estimate a value. For now, I'll remove it.

I've tested all the profiles, and all the data is present. You might be seeing an empty graph (the default endpoint of the Web UI) because, for some (for example, heap), the graph is only available on "snapshots." You can try them by running `teleport debug profile -s 0 heap`.

Reply:
It is tricky to define a default value for some profiles. For example, the heap profile collects information about GC and allocations. There are some cases where the sampling is too small to capture any information. (Note that when we do not provide seconds, or it is zero, the [profiler performs a GC](https://cs.opensource.google/go/go/+/refs/tags/go1.22.2:src/net/http/pprof/pprof.go;l=258), so there will always be data available.)

Setting the default `seconds` value won't work with the current default profiles, as `profile` only works with the delta profile.

What do you think about setting the default `seconds` value to zero and removing `profile` from the default profiles? This way, we'll always have data when running the `teleport debug profile` without any flags.

Reply:
I've updated the default value of `seconds` to `0` (snapshot) but added validation on `profile` and `trace` where we set a default value to avoid returning an error for those two (they require `seconds > 0`). This way, the default profiles will work correctly and always present data.